### PR TITLE
Add forecast module

### PR DESCRIPTION
**Pull request by fehiepsi**
Address #540.

The `GlobalTrendModel` is different from the current forecasting tutorial at the following points:
+ merge train and forecast into a single model
+ have a regression component
+ support all models from Rlgt library
+ the interface is similar to Pyro

TODO:
- [x] add lgt model
- [x] add sgt model
- [x] add sgt2 model
- [x] support forecasting with those models
- [x] ~vectorize the models~ not necessary because we will mainly use cpu
- [x] finalize the api
- [ ] add tests on toy datas
- [x] add docs
- [x] evaluate and backtesting ?

### CODE REVIEW

**Comment by fehiepsi** [comment]
@fritzo The current api can be seen at [this notebook](https://github.com/pyro-ppl/numpyro/blob/36457e8e15cd898ded44a2a8d5fd038d429606a1/numpyro/contrib/forecast/lynx.ipynb). WDYT? Do you have any suggestions for additional features that need to be implemented? (e.g. do we need `backtest` for MCMC?)

**Comment by fritzo** [comment]
@fehiepsi The interface looks good, thanks for staying close to the `pyro.contrib.forecast` interface. I think the `backtest()` function is not necessary in the first PR, but I do think it is important for performing experiments and for model exploration. Maybe you could implement it in a follow-up PR?

File: numpyro/contrib/forecast/gt.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    "pow_season_alpha": 1,
+    "pow_season_beta": 1,
+    "min_sigma": 1e-10,
+    "min_val": 1e-30,
+    "max_val": 1e38,
+}
+
+
+class GlobalTrendModel:
+    def __init__(self,
```
Comment (fritzo) [inline numpyro/contrib/forecast/gt.py]: nit: now that we're python 3 only, I've have started using the `*,` separator to force kwargs. Does this work with NumPyro? If so you could define this as
```py
class GlobalTrendModel:
    def __init__(self, *,
                 seasonality=1,
                 ...
                 hyperparameters={}):
```
This allows us to freely reorder parameters in the future, while maintaining backwards compatibility.

File: numpyro/contrib/forecast/gt.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+            powx = numpyro.sample("powx", dist.Uniform())
+            omega = sigma * exp_val ** powx + offset_sigma
+
+        if covariates.shape[0] == N:  # training
+            numpyro.sample("y", dist.StudentT(nu, exp_val, omega), obs=data[1:])
+        else:  # forecasting
+            numpyro.sample("y", dist.StudentT(nu, exp_val, omega))
+
+
+# src: https://github.com/cbergmeir/Rlgt/blob/master/Rlgt/src/stan_files/LGT.stan
```
Comment (fritzo) [inline numpyro/contrib/forecast/gt.py]: The source file is GPL licensed. Can you find a non-GPL source for a similar model?

File: numpyro/contrib/forecast/gt.py (hunk 1 above)
Comment (fehiepsi) [reply numpyro/contrib/forecast/gt.py]: Thanks, I'll do it and add a backtest function too.

File: numpyro/contrib/forecast/gt.py (hunk 2 above)
Comment (fehiepsi) [reply numpyro/contrib/forecast/gt.py]: Hmm, this is an important point... I'll learn more about the licence to make changes accordingly.

Comment (fehiepsi) [reply numpyro/contrib/forecast/gt.py]: Looking like there is no alternative unless we add a GPL license. I'll close this PR and think about other models which I can implement myself...

**Pull request closed without merging**